modules/core/src/arithm.cpp

```diff
@@ -1233,7 +1233,7 @@ void cv::compare(InputArray _src1, InputArray _src2, OutputArray _dst, int op)
     CV_Assert( op == CMP_LT || op == CMP_LE || op == CMP_EQ ||
                op == CMP_NE || op == CMP_GE || op == CMP_GT );
 
-    if(_src1.empty() && _src2.empty())
+    if(_src1.empty() || _src2.empty())
```

Comment:
There is no sense to compare some non-empty Mat with empty other Mat.
```
if(_src1.empty() && _src2.empty())
{
    dst.release();
    return;
}
CV_Assert(!_src1.empty() && !_src2.empty());
```

I would just raise an error if some Mat is empty (see #8300).